lib/amazonka-core/src/Amazonka/Error.hs

```diff
@@ -135,22 +137,23 @@ parseXMLError ::
   [Header] ->
   ByteStringLazy ->
   Error
-parseXMLError a s h bs = decodeError a s h bs (decodeXML bs >>= go)
+parseXMLError a s h bs = decodeError a s h bs (go <$> decodeXML bs)
   where
     go x =
       serviceError a s h
-        <$> code x
-        <*> may' (firstElement "Message" x)
-        <*> may' (firstElement "RequestId" x <|> firstElement "RequestID" x)
+      (code x)
+      (may' (firstElement "Message" x))
+      (may' (firstElement "RequestId" x) <|> may' (firstElement "RequestID" x))
 
-    code x =
-      Just <$> (firstElement "Code" x >>= parseXML)
-        <|> return root
+    code x
+      | Right y <- firstElement "Code" x >>= parseXML = y
+      | otherwise = root
 
     root = newErrorCode <$> rootElementName bs
 
-    may' (Left _) = pure Nothing
-    may' (Right x) = Just <$> parseXML x
+    may' x
+      | Right y <- x >>= parseXML = Just y
+      | otherwise = Nothing
```

Comment:
Similarly:
```suggestion
    may' x = either Nothing Just $ parseXML =<< x
```
(Unless there's a better tool in `Amazonka.Prelude` or something.)